server/src/main/java/org/opensearch/tasks/TaskResourceTrackingService.java

```diff
@@ -51,6 +52,7 @@
 /**
  * Service that helps track resource usage of tasks running on a node.
  */
+@PublicApi(since = "2.16.0")
```

Comment:
I understand why you are marking this public in that it is now exposed publicly from ClusterService.  Is this the right level of visibility for this service?  Can we instead expose only the required functionality from ClusterService instead the whole of `TaskResourceTrackingService`?  and/or thinking this maybe should be a separate interface for plugins not exposed through ClusterService?

Reply:
We can do a wrapper of `refreshResourceStats` in `clusterservice`, but one argument could be, does it belong to cluster service? In the right level of encapsulation, task level resource usages related operations should only belong to `TaskResourceTrackingService`. But I agree making the whole service public is also risky. I'm open to suggestions.

Reply:
Hi @reta ! thanks for the input. Currently the TaskResourceTracking Service only refreshes task usages when a task ends. But in our case we want to get the resource usage in a SearchOperationsListener, which will be triggered before a task finishes. 

Let me think about this more. Instead of exposing the `TaskResourceTrackingSevice`, does it make sense to you If we can do a usage refresh in core before the listeners are called?

Reply:
Thanks @ansjcy 

> thanks for the input. Currently the TaskResourceTrackingService only refreshes task usages when a task ends. 

It does not sound right, the task is still ongoing so it is usage won't be correct.

> Let me think about this more. Instead of exposing the TaskResourceTrackingSevice, does it make sense to you If we can do a usage refresh in core before the listeners are called (in this function:

The logical point (at least to me) of capturing task usages seems to be the moment task ends. It looks to me you are trying to chime in somewhere in between (while task is still executing), thanks does not look like the correct way.

Reply:
Yes, the task is still ongoing but the most heavy tasks (shardSearch tasks) are already done. Say a task is using 10000 ns CPU time, In this way we can get the best estimation with ~9500 ns and it can be used by the listener in query insights plugin to infer to top queries. What query insights care about are the top n queries ranking with high query level resource usages, it doesn't have to be 100% accurate.

Also, this doesn't affect anything other than query insights since we are still refreshing the usage when the task ends anyway, other downstream components like search back pressure will still get the accurate resource usages info. 

What do you think? Or let me know if you have any suggestions on the approaches.

Reply:
Another possible solution is, we add a `Supplier` in search request context to call taskresourcetracking service to do the refresh. In plugins, we just need to invoke that supplier.

Reply:
Ah I do remember this discussion, thanks @ansjcy . So I think your idea with adding dedicated `SearchRequestOperationsListener` is viable: the `onRequestEnd` callback has reference to the `SearchTask` which could ask for tracking refresh. Sounds like a way to go?

Reply:
yes :) ! however `SearchTask` is not able to refresh itself, we could either add a supplier into SearchTask to allow it to refresh itself, or add the supplier into the SearchRequestContext, similar to what we did here: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/action/search/TransportSearchAction.java#L441

Reply:
Hm ..the `onRequestEnd` already has `SearchRequestContext` argument, which exposes the `getTaskResourceUsageSupplier()`, or I am missing something?

Reply:
That's a supplier to get the "current" resource usage for tasks executed on data nodes, and we want to get the "refreshed" resource usages. So we still need another supplier to do the refresh for the current thread.

Reply:
May be I was thinking in a bit different direction: we could add the `SearchRequestOperationsListener` impl (that could reference `TaskResourceTrackingService`, both are internal to core), and  `onRequestEnd`  could call `TaskResourceTrackingService::refreshResourceStats` (since we have access to task)

Reply:
> we could add the SearchRequestOperationsListener impl (that could reference TaskResourceTrackingService,

hmm, are you suggesting we add a new `SearchRequestOperationsListener` class with `TaskResourceTrackingService` as a field, and query insights listener can inherit that class instead of the original `SearchRequestOperationsListener` class?

Reply:
No, we add per-request listener instance in the `TransportSearchAction::executeRequest`. The query insights plugin has nothing to do with it at this moment of time, but we capture the search task resource usage upon request completion, so the tracking data becomes available to everyone (including the query insights plugin). Does it make sense?

Reply:
I see. It make sense to me! Let me update the PR with this.

Reply:
@reta one concern on this approach is, since query insights is also using SearchRequestOperationsListener as the callback, but right now we add per request listeners after adding query insights listener: https://github.com/opensearch-project/OpenSearch/blob/8ae728c5610900888449af2eae2ef8b23d664538/server/src/main/java/org/opensearch/action/search/SearchRequestOperationsCompositeListenerFactory.java#L73
We might need to change the sequence in the above function to
```

final List<SearchRequestOperationsListener> searchListenersList = Stream.concat(
      Arrays.stream(perRequestListeners)
        ),
searchRequestListenersList.stream()
```

Reply:
> @reta one concern on this approach is, since query insights is also using SearchRequestOperationsListener as the callback,

@ansjcy wait, `SearchRequestOperationsListener` is `@InternalApi`, how come the plugin uses it?

Reply:
@reta , well, it should be a publicAPI. Before we start implementing query insights, one of the prerequisites was able to use this `SearchRequestOperationsListener ` in plugins so that plugins like query insights will be able to get execution information and context related to the request. [This issue](https://github.com/opensearch-project/OpenSearch/issues/11520) along with [this PR](https://github.com/opensearch-project/OpenSearch/pull/11526) was trying to solve this by refactoring how the listeners are created and managed. 

We didn't change the api annotation from `@InternalApi` to `publicAPI` at that time, however we didn't get any build failures or other check failures. Do we want to revisit how those checks are done? Meanwhile we might need to change it to public as well.

Reply:
> @reta , well, it should be a publicAPI.

@ansjcy I think this needs justification and some work, it is not about changing annotation - there are tons of other abstractions that will require to be exposed as public. 

> We didn't change the api annotation from @InternalApi to publicAPI at that time, however we didn't get any build failures or other check failures. Do we want to revisit how those checks are done? Meanwhile we might need to change it to public as well.

We cannot enforce the usage of `@InternalApi` since the plugins / extensions / etc could use whatever build system or scripts. The `@InternalApi` hints that there is no guarantee about anything that such APIs provide - it could entirely disappear any time.

Reply:
> No, we add per-request listener instance in the TransportSearchAction::executeRequest. The query insights plugin has nothing to do with it at this moment of time, but we capture the search task resource usage upon request completion, so the tracking data becomes available to everyone (including the query insights plugin). Does it make sense?

@reta We are currently working on a [PR](https://github.com/opensearch-project/OpenSearch/pull/14782/files) based on the above discussion.

For the last point you made, what other justification and work is required to make the API public? We are trying to get all the query insights changes in 2.16 and this is the only PR that is dangling currently. Want to make sure we reach a path forward. Please let us know your suggestions. In the meantime will finalize the above [draft PR](https://github.com/opensearch-project/OpenSearch/pull/14782/files) if there are not concerns with this approach?

Reply:
Thanks @deshsidd 

> For the last point you made, what other justification and work is required to make the API public?

I did not design the original APIs, you may ask the contributor if he has any concerns. On the second point, if you need to make it public, apply the @PublicApi annotation accordingly.

Reply:
Thanks @reta!
cc @buddharajusahil @sgup432 @dzane17 @jainankitk Please let us know your thoughts since you all have contributed to the file. 

@reta Looks like you had [initially reduced the visibility of the API](https://github.com/opensearch-project/OpenSearch/pull/11242/files).

For now I am going to continue with the approach that Reta and Chenyang had discussed above and work on the [following PR](https://github.com/opensearch-project/OpenSearch/pull/14782/files). Will also make the SearchRequestOperationsListener @PublicApi as part of these changes.

Reply:
> @reta Looks like you had [initially reduced the visibility of the API](https://github.com/opensearch-project/OpenSearch/pull/11242/files).

@deshsidd yes, you will understand why once try to apply `@PublicApi` to it :-) : it pulls a pile of dependencies with it .... (I am very doubtful it was designed as being public in the first place).

Reply:
Understood.
@reta and others - Please take a look : https://github.com/opensearch-project/OpenSearch/pull/14832 and let me know your thoughts.